### PR TITLE
new input filter : nibble_swap

### PR DESCRIPTION
**Pull request by fenugrec**
Simple, exchange upper and lower halves of each byte, e.g. "A6" -> "6A".

Includes a simple test script.

### CODE REVIEW

File: srecord/input/filter/nibble_swap.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#ifndef SRECORD_INPUT_FILTER_NIBBLE_SWAP_H
+#define SRECORD_INPUT_FILTER_NIBBLE_SWAP_H
+
+#include <srecord/input/filter.h>
+
+namespace srecord {
+
+/**
+  * The srecord::input_filter_nibble_swap class is used to represent the input state
+  * of a filter which reverses the order of bits in each data byte.
```
Comment (sierrafoxtrot) [inline srecord/input/filter/nibble_swap.h:27]: Suggest 

`The srecord::input_filter_nibble_swap class is used to represent the input state
 of a filter which exchanges upper and lower nibble of each byte, e.g. "0xA6" => "0x6A"`

File: srecord/input/filter/nibble_swap.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
+//
+// srecord filter that exchanges upper and lower nibble of each byte, e.g. "0xA6" => "0x6A"
```
Comment (sierrafoxtrot) [inline srecord/input/filter/nibble_swap.h:2]: Please replace with a copyright assertion to a legal name (rather than username/handle). See another similar files for examples. 

`// srecord - Manipulate EPROM load files`
`// Copyright (C) 2023 Firstname Lastname`

File: srecord/input/filter/nibble_swap.cc [hunk 3]
```diff
@@ -0,0 +1,50 @@
+//
+// srecord filter that exchanges upper and lower nibble of each byte, e.g. "0xA6" => "0x6A"
```
Comment (sierrafoxtrot) [inline srecord/input/filter/nibble_swap.cc:2]: Please replace this line with a copyright assertion to a legal name (rather than username/handle). See another similar files for examples.

`// srecord - Manipulate EPROM load files`
`// Copyright (C) 2023 Firstname Lastname`

File: test/02/t0269a.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
+#
+#       srecord - manipulate eprom load files
+#       Copyright (C) 2023 fenugrec
```
Comment (sierrafoxtrot) [inline test/02/t0269a.sh:4]: Please give a legal name for copyright assignment.

File: doc/man1/srec_input.1 [hunk 5]
```diff
@@ -540,6 +540,10 @@ of 4 and 8 bytes, the default is 2 bytes.  (Widths in excess of 8 are
 assumed to be number of bits.)
 It is not possible to swap non\[hy]power\[hy]of\[hy]two addresses.
 To change the alignment, use the offset filter before and after.
+.\"             nibble swap
+.TP 8n
+\fB\-Nibble_Swap\fP [ \f[I]width\fP ]
```
Comment (sierrafoxtrot) [inline doc/man1/srec_input.1:545]: The width argument doesn't appear to be used. Perhaps a copy-paste error? 

Suggest : `\fB\-Nibble_Swap\fP `

Otherwise, will need to be handled in `srecord/arglex/tool/input.cc`

**Review by sierrafoxtrot — CHANGES_REQUESTED** [review]
This is looking great @fenugrec! A few minor(I hope) comments and some small line length items from the CI lint tests. 

Would you please also update `doc/etc/README.man` (around line 318) to add the filter

File: srecord/input/filter/nibble_swap.h (hunk 1 above)
Comment (fenugrec) [reply srecord/input/filter/nibble_swap.h:27]: Good catch, some copypasta leftovers there. will fix

File: doc/man1/srec_input.1 (hunk 5 above)
Comment (fenugrec) [reply doc/man1/srec_input.1:545]: more copypaste, will fix

**Comment by fenugrec** [comment]
Thanks for reviewing. Will update this list as I go

- [ ] line length items from the CI lint tests.
- [ ]   doc/etc/README.man (around line 318) to add the filter
- [ ] copyright
- [ ] other inline fixes

**Comment by fenugrec** [comment]
Ok, the megalint long-line errors should be fixed.

About the copyright notice : if I cannot use my handle there, I assigned it to "srecord project". If that is not satisfactory then I will copy the notice from the bit_reverse (or byte_swap) filters which I mostly duplicated except for one line.

**Comment by sierrafoxtrot** [comment]
> Ok, the megalint long-line errors should be fixed.
> 
> About the copyright notice : if I cannot use my handle there, I assigned it to "srecord project". If that is not satisfactory then I will copy the notice from the bit_reverse (or byte_swap) filters which I mostly duplicated except for one line.
> 
> Note : there's already a `// Copyright (C) 2019 fenugrec` line in input/file/hp64.* ...

> Ok, the megalint long-line errors should be fixed.

Fantastic. Looks good. 

> About the copyright notice : if I cannot use my handle there, I assigned it to "srecord project". If that is not satisfactory then I will copy the notice from the bit_reverse (or byte_swap) filters which I mostly duplicated except for one line.

That is probably a solid approach although people may get a little confused about files with circa 2010 copyright being created in 2023. If they are different, we could have an issue as Peter can no longer assert copyright and his estate has no legal entity. But if this is your preferred approach, I think on balance, we can go with it. 

> Note : there's already a `// Copyright (C) 2019 fenugrec` line in input/file/hp64.* ...

I spotted the existing one too and that was entirely my fault for letting it through. The problem is that the GPL relies on assertion/defence of copyright which in many if not all jurisdictions can only be made by a legal entity ie a person or business. While I appreciate the intent with "srecord project", it isn't a legal entity either. 

I completely support your preference to keep your legal name out of the code base so the best I can offer is if you assign it to me. I don't like the idea of claiming credit for your work (history still shows it comes from you) but unless you have a corporate entity that could hold the copyright, it's the best we have available. Perhaps something like the following?
`// Copyright (C) 2019 Scott Finneran (authored by fenugrec & copyright assigned)`

Now I remember why I chose engineering over law :-)

File: srecord/input/filter/nibble_swap.cc [hunk 6]
```diff
@@ -0,0 +1,54 @@
+//
+// Copyright (c) 2023 srecord project
```
Comment (sierrafoxtrot) [inline srecord/input/filter/nibble_swap.cc:2]: Please see comments around copyright assignment to some form of legal entity

File: srecord/input/filter/nibble_swap.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) 2023 srecord project
```
Comment (sierrafoxtrot) [inline srecord/input/filter/nibble_swap.h:2]: Please see comments around copyright assignment to some form of legal entity

File: test/02/t0269a.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
+#
+#       srecord - manipulate eprom load files
+#       Copyright (C) 2023 srecord project
```
Comment (sierrafoxtrot) [inline test/02/t0269a.sh:4]: Please see comments around copyright assignment to some form of legal entity

**Review by sierrafoxtrot — CHANGES_REQUESTED** [review]

**Comment by fenugrec** [comment]
> we could have an issue as Peter can no longer assert copyright and his estate has no legal entity

Good point.

> The problem is that the GPL relies on assertion/defence of copyright which in many if not all jurisdictions can only be made by a legal entity

Yeah, getting fast into uncharted waters for me, and of course varies according to local law, but there is definitely some precedent for asserting copyright under a pseudonym (not a new thing in the publishing world). What few references I've found all  recommend consulting lawyers , obviously :
- Berne convention:  https://www.wipo.int/treaties/en/ip/berne/summary_berne.html " In the case of anonymous or pseudonymous works, the term of protection expires 50 years after the work has been lawfully made available to the public,"
- US, https://web.archive.org/web/20210428165628if_/https://www.copyright.gov/fls/fl101.pdf "An author of a copyrighted work can use a pseudonym or pen name. [...]"
- Canada, https://laws-lois.justice.gc.ca/eng/acts/C-42/page-3.html?txthl=pseudonym#s-14.1 "The author of a work has [...] the right, [...] to be associated with the work as its author by name or under a pseudonym and the right to remain anonymous. "
- UK is weird, https://copyrightservice.co.uk/copyright/p03_copyright_notices "Although it may not be technically correct (it does not state the name of the legal entity that is the copyright owner), it is very common for an identifiable pseudonym or trading name to be used in the copyright notice"

I think using "srecord project' would probably be valid in many jurisdictions, I've seen various wording like "works produced in collaboration / in a group or team", but that's another story.

As I understand since I published this PR (and the hp64k one) under GPL under a pseudonym, the copyright enforceability is my problem - it says GPL and nothing else, so you are free to include it in srecord under the GPL terms.  You don't inherit the responsibility of protecting my copyright, since it's mine ?

Anyway, as maintainer, it's entirely up to you how to proceed; if it's a problem then for this PR at least I have no objection to you claiming copyright on it. It's trivial code that you could have produced in less time than it took me to write this reply -- I hardly even feel I deserve to have my own name there !

> Now I remember why I chose engineering over law :-)

Yeah, no joke. The only laws I want to deal with are the laws of physics.

**Comment by sierrafoxtrot** [comment]
>     * UK is weird, https://copyrightservice.co.uk/copyright/p03_copyright_notices "Although it may not be technically correct (it does not state the name of the legal entity that is the copyright owner), it is very common for an identifiable pseudonym or trading name to be used in the copyright notice"
Yep and we have some similar conventions here in Australia. Hence my move to keeping it simple. 

> Anyway, as maintainer, it's entirely up to you how to proceed; if it's a problem then for this PR at least I have no objection to you claiming copyright on it. It's trivial code that you could have produced in less time than it took me to write this reply -- I hardly even feel I deserve to have my own name there !

Mate, you took the time to contribute and even consider my comments above seriously which I greatly appreciate. In the interests of taking a simple approach that I can apply consistently, would you mind making the updates to something like 
// Copyright (C) 2023 Scott Finneran (authored by fenugrec & copyright assigned)

or perhaps 
//  Authored by fenugrec with copyright assigned to Scott Finneran 
// Copyright (C) 2023 Scott Finneran

> > Now I remember why I chose engineering over law :-)
> 
> Yeah, no joke. The only laws I want to deal with are the laws of physics.
LOL, much more reasonable set of rules!

**Comment by fenugrec** [comment]
> Yep and we have some similar conventions here in Australia. Hence my move to keeping it simple.

Right. Thought about it some more, and still, I can't see how this would ever be a problem. Even if you or some future maintainer wants to relicense everything, whether there's a real or fake name or an obvious pseudonym in the notice won't change much, as I would still need to be contacted for permission. If someone uses parts of my code and doesn't observe GPL requirements, that's not the project's problem to deal with but entirely mine. Can you think of a scenario where it would cause issues ?

**Comment by sierrafoxtrot** [comment]
> > Yep and we have some similar conventions here in Australia. Hence my move to keeping it simple.
> 
> Right. Thought about it some more, and still, I can't see how this would ever be a problem. Even if you or some future maintainer wants to relicense everything, whether there's a real or fake name or an obvious pseudonym in the notice won't change much, as I would still need to be contacted for permission. If someone uses parts of my code and doesn't observe GPL requirements, that's not the project's problem to deal with but entirely mine. Can you think of a scenario where it would cause issues ?

The GPL essentially applies project-wide. The license isn't per-file and as such, weakening assertion/defense on any file weakens it for the whole project. This is _one of_ the reasons that entities such as the GNU project and the Software Freedom Conservancy insist on copyright assignment across the entire codebase.

**Comment by sierrafoxtrot** [comment]
> > > Yep and we have some similar conventions here in Australia. Hence my move to keeping it simple.
> > 
> > 
> > Right. Thought about it some more, and still, I can't see how this would ever be a problem. Even if you or some future maintainer wants to relicense everything, whether there's a real or fake name or an obvious pseudonym in the notice won't change much, as I would still need to be contacted for permission. If someone uses parts of my code and doesn't observe GPL requirements, that's not the project's problem to deal with but entirely mine. Can you think of a scenario where it would cause issues ?
> 
I gave it some more thought and sought some more advice. 

Regarding copyright assertion, a handle is essentially a pseudonym (aka pen-name) and this doesn't undermine copyright assertion as long as we can trace it back to a real identity. 

Your commits are attributed to an email address and possibly even signed giving traceability back to a contact point. It's an email redirect via Sourceforge but better than nothing. I may end up turning on "commits must be signed" at some point. 

Bottom line is that on balance, I do think we'll be ok. Let's move forward as-is on the copyright assignment. Thanks for helping me think this through.

**Comment by sierrafoxtrot** [comment]
When you get a moment to change the copyright assignments to fenugrec, I can merge this in.

**Comment by sierrafoxtrot** [comment]
> > When you get a moment to change the copyright assignments to fenugrec, I can merge this in.
> 
> Oh I thought I alread had, did I miss one ?

Sorry, my mistake, I mucked something up pulling latest. All good!

**Review by sierrafoxtrot — APPROVED** [review]

**Comment by fenugrec** [comment]
awesome. thanks
